File: July/Day25/maximum-unique-subarray-sum-after-deletion.js
// HashMap Approach  ::  Time: O(N) ::  Space: O(N)
class Solution {
    maxSum(nums) {
        const seen = new Set();
        let maxSum = -Infinity;
        let isFullNegative = true;

        for (let num of nums) {
            if (num >= 0) {
                isFullNegative = false;
                maxSum = 0;
            }
        }

        for (let num of nums) {
            if (isFullNegative) {
                maxSum = Math.max(maxSum, num);
            } else if (!seen.has(num) && num > 0) {
                seen.add(num);
                maxSum += num;
            }
        }

        return maxSum;
    }
}




// Sorting Approach  ::  Time: O(Nlogn) + O(N) ::  Space: O(1)
class Solution {
    maxSum(nums) {
        nums.sort((a, b) => a - b);
        let maxSum = -Infinity;
        let prev = -1;

        for (let num of nums) {
            if (num <= 0) {
                maxSum = Math.max(maxSum, num);
            } else if (num !== prev) {
                if (maxSum < 0) maxSum = 0;
                maxSum += num;
                prev = num;
            }
        }

        return maxSum;
    }
}

File: July/Day25/maximum-unique-subarray-sum-after-deletion.py
# HashMap Approach  ::  Time: O(N) ::  Space: O(N)
class Solution:
    def maxSum(self, nums: List[int]) -> int:
        freq =Counter()
        maximumSum = -inf
        isFullNegative = True
        for num in nums:
            if num >= 0:
                isFullNegative = False
                maximumSum = 0


        for num in nums:
            if isFullNegative:
                maximumSum = max(num, maximumSum)
            elif num not in freq and num > 0:
                freq[num] = 1
                maximumSum+=num
            
        
        return maximumSum



# Sorting Approach  ::  Time: O(Nlogn) + O(N) ::  Space: O(1)
class Solution:
    def maxSum(self, nums: List[int]) -> int:
        nums.sort()
        prev = maximumSum =-inf


        for num in nums:
            if num <= 0:
                maximumSum = max(num, maximumSum)
            elif prev != num:
                if maximumSum < 0:
                    maximumSum = 0
                prev = num
                maximumSum+= num
        
        return maximumSum